### PR TITLE
Add Span<T> Base64 conversion APIs that support UTF-8

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/dotnet/corefx/issues/24568 (part of https://github.com/dotnet/corefx/issues/24174).

```C#
namespace System.Buffers.Text {
    public static class Base64 {
        public static OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Span<byte> bytes, out int consumed, out int written, bool isFinalBlock=true);
        public static OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, out int written);
        public static OperationStatus EncodeToUtf8(ReadOnlySpan<byte> bytes, Span<byte> utf8, out int consumed, out int written, bool isFinalBlock=true);
        public static OperationStatus EncodeToUtf8InPlace(Span<byte> buffer, int dataLength, out int written);
        public static int GetMaxDecodedFromUtf8Length(int length);
        public static int GetMaxEncodedToUtf8Length(int length);
    }
}

namespace System.Buffers {
    public enum OperationStatus {
        DestinationTooSmall = 1,
        Done = 0,
        InvalidData = 3,
        NeedMoreData = 2,
    }
}
```

**Note:**
OperationStatus (previously TransformationStatus) can be updated base on API review and doesn't need to block this PR (https://github.com/dotnet/corefx/issues/22412).

**Code Coverage:**
100% line and branch coverage
image

**Performance Test Results (duration):**

Test   Name | AVERAGE
-- | --
Base64Decode(numberOfBytes: 10) | 0.019
Base64Decode(numberOfBytes: 100) | 0.110
Base64Decode(numberOfBytes: 1000) | 0.950
Base64Decode(numberOfBytes: 1000000) | 954.361
Base64DecodeBaseline(numberOfBytes: 10) | 0.072
Base64DecodeBaseline(numberOfBytes: 100) | 0.420
Base64DecodeBaseline(numberOfBytes: 1000) | 3.819
Base64DecodeBaseline(numberOfBytes:   1000000) | 8529.559
Base64DecodeDetinationTooSmall(numberOfBytes:   10) | 0.018
Base64DecodeDetinationTooSmall(numberOfBytes:   100) | 0.115
Base64DecodeDetinationTooSmall(numberOfBytes:   1000) | 0.948
Base64DecodeDetinationTooSmall(numberOfBytes:   1000000) | 954.096
Base64DecodeInPlace(numberOfBytes: 10) | 0.020
Base64DecodeInPlace(numberOfBytes: 100) | 0.113
Base64DecodeInPlace(numberOfBytes: 1000) | 0.981
Base64DecodeInPlace(numberOfBytes:   1000000) | 1023.155
Base64DecodeInPlaceOnce(numberOfBytes:   1000000) | 0.950
Base64Encode(numberOfBytes: 10) | 0.020
Base64Encode(numberOfBytes: 100) | 0.129
Base64Encode(numberOfBytes: 1000) | 1.158
Base64Encode(numberOfBytes: 1000000) | 1170.335
Base64EncodeBaseline(numberOfBytes: 10) | 0.029
Base64EncodeBaseline(numberOfBytes: 100) | 0.145
Base64EncodeBaseline(numberOfBytes: 1000) | 1.351
Base64EncodeBaseline(numberOfBytes:   1000000) | 1303.068
Base64EncodeDestinationTooSmall(numberOfBytes:   10) | 0.018
Base64EncodeDestinationTooSmall(numberOfBytes:   100) | 0.128
Base64EncodeDestinationTooSmall(numberOfBytes:   1000) | 1.162
Base64EncodeDestinationTooSmall(numberOfBytes:   1000000) | 1163.685
Base64EncodeInPlace(numberOfBytes: 10) | 0.023
Base64EncodeInPlace(numberOfBytes: 100) | 0.133
Base64EncodeInPlace(numberOfBytes: 1000) | 1.171
Base64EncodeInPlace(numberOfBytes:   1000000) | 1247.834
Base64EncodeInPlaceOnce(numberOfBytes:   1000000) | 1.176

<details> 
<summary>Out-dated performance results 1</summary>

Test   Name | AVERAGE
-- | --
Base64Decode(numberOfBytes: 10) | 0.018
Base64Decode(numberOfBytes: 100) | 0.113
Base64Decode(numberOfBytes: 1000) | 1.005
Base64Decode(numberOfBytes: 1000000) | 1011.903
Base64DecodeBaseline(numberOfBytes: 10) | 0.083
Base64DecodeBaseline(numberOfBytes: 100) | 0.452
Base64DecodeBaseline(numberOfBytes: 1000) | 4.077
Base64DecodeBaseline(numberOfBytes:   1000000) | 8977.842
Base64DecodeInPlace(numberOfBytes: 10) | 0.022
Base64DecodeInPlace(numberOfBytes: 100) | 0.114
Base64DecodeInPlace(numberOfBytes: 1000) | 0.983
Base64DecodeInPlace(numberOfBytes:   1000000) | 1055.159
Base64DecodeInPlaceOnce(numberOfBytes:   1000000) | 0.985
Base64Encode(numberOfBytes: 10) | 0.024
Base64Encode(numberOfBytes: 100) | 0.171
Base64Encode(numberOfBytes: 1000) | 1.631
Base64Encode(numberOfBytes: 1000000) | 1659.814
Base64EncodeBaseline(numberOfBytes: 10) | 0.034
Base64EncodeBaseline(numberOfBytes: 100) | 0.149
Base64EncodeBaseline(numberOfBytes: 1000) | 1.342
Base64EncodeBaseline(numberOfBytes:   1000000) | 1340.315
Base64EncodeInPlace(numberOfBytes: 10) | 0.068
Base64EncodeInPlace(numberOfBytes: 100) | 0.406
Base64EncodeInPlace(numberOfBytes: 1000) | 3.782
Base64EncodeInPlace(numberOfBytes:   1000000) | 3983.673
Base64EncodeInPlaceOnce(numberOfBytes:   1000000) | 3.854
</details> 

<details> 
<summary>Out-dated performance results 2</summary>

Test   Name | AVERAGE
-- | --
Base64Decode(numberOfBytes: 10) | 0.019
Base64Decode(numberOfBytes: 100) | 0.118
Base64Decode(numberOfBytes: 1000) | 1.001
Base64Decode(numberOfBytes: 1000000) | 1000.071
Base64DecodeBaseline(numberOfBytes: 10) | 0.075
Base64DecodeBaseline(numberOfBytes: 100) | 0.428
Base64DecodeBaseline(numberOfBytes: 1000) | 3.872
Base64DecodeBaseline(numberOfBytes:   1000000) | 8725.017
Base64DecodeInPlace(numberOfBytes: 10) | 0.022
Base64DecodeInPlace(numberOfBytes: 100) | 0.117
Base64DecodeInPlace(numberOfBytes: 1000) | 0.986
Base64DecodeInPlace(numberOfBytes:   1000000) | 1068.851
Base64DecodeInPlaceOnce(numberOfBytes:   1000000) | 0.993
Base64Encode(numberOfBytes: 10) | 0.023
Base64Encode(numberOfBytes: 100) | 0.159
Base64Encode(numberOfBytes: 1000) | 1.437
Base64Encode(numberOfBytes: 1000000) | 1473.720
Base64EncodeBaseline(numberOfBytes: 10) | 0.036
Base64EncodeBaseline(numberOfBytes: 100) | 0.149
Base64EncodeBaseline(numberOfBytes: 1000) | 1.602
Base64EncodeBaseline(numberOfBytes:   1000000) | 1357.203
Base64EncodeInPlace(numberOfBytes: 10) | 0.027
Base64EncodeInPlace(numberOfBytes: 100) | 0.157
Base64EncodeInPlace(numberOfBytes: 1000) | 1.399
Base64EncodeInPlace(numberOfBytes:   1000000) | 1466.058
Base64EncodeInPlaceOnce(numberOfBytes:   1000000) | 1.402
</details> 

<details> 
<summary>Out-dated performance results 3</summary>

Test   Name | AVERAGE
-- | --
Base64Decode(numberOfBytes: 10) | 0.017
Base64Decode(numberOfBytes: 100) | 0.110
Base64Decode(numberOfBytes: 1000) | 0.984
Base64Decode(numberOfBytes: 1000000) | 1007.293
Base64DecodeBaseline(numberOfBytes: 10) | 0.069
Base64DecodeBaseline(numberOfBytes: 100) | 0.410
Base64DecodeBaseline(numberOfBytes: 1000) | 3.941
Base64DecodeBaseline(numberOfBytes:   1000000) | 8593.280
Base64DecodeInPlace(numberOfBytes: 10) | 0.021
Base64DecodeInPlace(numberOfBytes: 100) | 0.116
Base64DecodeInPlace(numberOfBytes: 1000) | 1.008
Base64DecodeInPlace(numberOfBytes:   1000000) | 1045.012
Base64DecodeInPlaceOnce(numberOfBytes:   1000000) | 0.955
Base64Encode(numberOfBytes: 10) | 0.020
Base64Encode(numberOfBytes: 100) | 0.132
Base64Encode(numberOfBytes: 1000) | 1.585
Base64Encode(numberOfBytes: 1000000) | 1169.956
Base64EncodeBaseline(numberOfBytes: 10) | 0.029
Base64EncodeBaseline(numberOfBytes: 100) | 0.153
Base64EncodeBaseline(numberOfBytes: 1000) | 1.293
Base64EncodeBaseline(numberOfBytes:   1000000) | 1316.709
Base64EncodeInPlace(numberOfBytes: 10) | 0.023
Base64EncodeInPlace(numberOfBytes: 100) | 0.158
Base64EncodeInPlace(numberOfBytes: 1000) | 1.207
Base64EncodeInPlace(numberOfBytes:   1000000) | 1252.556
Base64EncodeInPlaceOnce(numberOfBytes:   1000000) | 1.166
</details> 

Encode is about  10% faster ~20% slower~ ~10% slower (for input larger than 100 bytes)~. Decode is at least 4x faster (even when compared to the new non-allocating variant TryFromBase64Chars). ~This is probably because the Convert.FromBase64CharArray allocates (approximately n bytes, where n is input size + around 24 byte overhead).~

**Edit 0:** Made encode a bit faster for the common case.
**Edit 1:** Compared against TryFromBase64Chars, there is still a large perf difference.
**Edit 2:** Added new performance numbers based on additional optimizations suggested by @jkotas.
**Edit 3:** Added performance tests for when destination buffer is too small.

cc @KrzysztofCwalina, @stephentoub, @GrabYourPitchforks, @jkotas, @dotnet/corefxlab-contrib

### CODE REVIEW

**Comment by jkotas** [comment]
> Convert.FromBase64CharArray allocates

It would be more fair to use the non-allocating variant.

**Comment by ahsonkhan** [comment]
> It would be more fair to use the non-allocating variant.

I couldn't find any non-allocating public API (https://msdn.microsoft.com/en-us/library/system.convert_methods(v=vs.110).aspx). Both of these return a `byte[]`:
- FromBase64CharArray
- FromBase64String

**Comment by jkotas** [comment]
The current .NET Core has TryFromBase64Chars

**Comment by ahsonkhan** [comment]
> The current .NET Core has TryFromBase64Chars

I updated the performance test. Although there wasn't much difference in the results.

**Comment by stephentoub** [comment]
> Decode is at least 4x faster (even when compared to the new non-allocating variant TryFromBase64Chars

Then can you also fix the baseline?  We shouldn't be happy about such a difference. We should be happy that we've found ways to improve the existing one and then do it. I can believe there could be a small difference due to fundamental things like outputting 1 byte vs 2 bytes per encoded unit, but it shouldn't be anywhere near that big.

Success isn't making the new thing faster. Success is making the new thing fast and then making the existing one as fast or as close to as fast as is possible.

File: src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+            ref byte destBytes = ref utf8.DangerousGetPinnableReference();
+
+            int srcLength = bytes.Length;
+            int destLength = utf8.Length;
+
+            int sourceIndex = 0;
+            int destIndex = 0;
+            int result = 0;
+
+            if (destLength >= GetMaxEncodedToUtf8Length(srcLength))
```
Comment (jkotas) [inline src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: This will throw when the source is too big. Shouldn't it rather return `OperationStatus.DestinationTooSmall` ?

File: src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        /// - DestinationTooSmall - if there is not enough space in the buffer beyond dataLength to fit the result of encoding the input
+        /// It does not return NeedMoreData since this method tramples the data in the buffer and hence can only be called once with all the data in the buffer.
+        /// It does not return InvalidData since that is not possible for base 64 encoding.</returns>
+        /// </summary> 
+        public static OperationStatus EncodeToUtf8InPlace(Span<byte> buffer, int dataLength, out int written)
+        {
+            var encodedLength = GetMaxEncodedToUtf8Length(dataLength);
+            if (buffer.Length < encodedLength) goto FalseExit;
+
+            var leftover = dataLength - dataLength / 3 * 3; // how many bytes after packs of 3
```
Comment (jkotas) [inline src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: Nit: It is not ok to use var here per corefx coding conventions.

File: src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+            int i0 = b0 >> 2;
+            r0 = s_encodingMap[i0];
+
+            int i1 = (b0 & 0x3) << 4 | (b1 >> 4);
+            r1 = s_encodingMap[i1];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int Encode(ref byte threeBytes)
+        {
```
Comment (jkotas) [inline src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: You should be able to use same set of tricks here as you are using for Decode: cache the map in a local upfront, avoid bound checks for it and save some binary ops by operating on the whole byte. Something like:

```
private static int Encode(ref byte threeBytes, ref byte encodingMap)
{
    int i = (threeBytes << 16) | (Unsafe.Add(ref threeBytes, 1) << 8) | Unsafe.Add(ref threeBytes, 2);

    int i0 = Unsafe.Add(ref encodingMap, i >> 18)
    int i1 = Unsafe.Add(ref encodingMap, (i >> 12) & 0x3F);
    int i2 = Unsafe.Add(ref encodingMap, (i >> 6) & 0x3F);
    int i3 = Unsafe.Add(ref encodingMap, i & 0x3F);

    return i0 | (i1 << 8) | (i2 << 16) | (i3 << 24);
}
```

File: src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+            73, 74, 75, 76, 77, 78, 79, 80,         //I..P
+            81, 82, 83, 84, 85, 86, 87, 88,         //Q..X
+            89, 90, 97, 98, 99, 100, 101, 102,      //Y..Z, a..f
+            103, 104, 105, 106, 107, 108, 109, 110, //g..n
+            111, 112, 113, 114, 115, 116, 117, 118, //o..v
+            119, 120, 121, 122, 48, 49, 50, 51,     //w..z, 0..3
+            52, 53, 54, 55, 56, 57, 43, 47          //4..9, +, /
+        };
+
+        const byte s_encodingPad = (byte)'='; // '=', for padding
```
Comment (jkotas) [inline src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: Nit: `s_` is not correct prefix for consts per corefx coding conventions

File: src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+            return OperationStatus.InvalidData;
+        }
+
+        /// <summary>
+        /// Returns the maximum length (in bytes) of the result if you were to deocde base 64 encoded text within a byte span of size "length".
+        /// </summary> 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int GetMaxDecodedFromUtf8Length(int length)
+        {
+            Debug.Assert(length >= 0);
```
Comment (KrzysztofCwalina) [inline src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs]: Do we like debug assert here? Should we return 0 or throw?

File: src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+        /// <summary>
+        /// Returns the maximum length (in bytes) of the result if you were to encode binary data within a byte span of size "length".
+        /// </summary>
+        /// <exception cref="System.OverflowException">
+        /// Thrown when the specified <paramref name="length"/> is larger than 1610612733 (since encode inflates the data by 4/3).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int GetMaxEncodedToUtf8Length(int length)
+        {
+            Debug.Assert(length >= 0);
```
Comment (KrzysztofCwalina) [inline src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: Same, i.e. I think we should throw for negative lengths.

Comment (bartonjs) [reply src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: Agreed.  Public members need to either accept their inputs (no assert, no check) or properly validate (check and throw).  Asserting on inputs is only valid for non-public members, to guard that the public member already did the validation.

Comment (GrabYourPitchforks) [reply src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: Do we care what exception type gets thrown? Personally I don't care if it's ArgumentException or OverflowException, or even if it's potentially both (negative vs. too large). It's not like the developer can catch and handle it anyway.

Comment (bartonjs) [reply src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: ArgumentOutOfRangeException seems to match the problem space to me.  (Well, really, argument out of domain (since it's an input); but that's just my math degree talking)

**Comment by ahsonkhan** [comment]
> Then can you also fix the baseline? We shouldn't be happy about such a difference. 

I agree. Filed an [issue](https://github.com/dotnet/corefx/issues/24897) to look into this. Ty.

File: src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: I considered that. The issue with DestinationTooSmall is the user will then expect to enlarge the output buffer and try again, and for any buffer size up to int.MaxValue, they won't succeed for the current input length.

However, if they could continue to slice the input as it is partially consumed, then they would eventually succeed (or they could get unmanaged memory using AllocHGlobal). So, I will update the behavior as per your suggestion.

File: src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs [hunk 7]
```diff
@@ -40,11 +40,13 @@ public static OperationStatus EncodeToUtf8(ReadOnlySpan<byte> bytes, Span<byte>
             int destIndex = 0;
             int result = 0;
 
-            if (destLength >= GetMaxEncodedToUtf8Length(srcLength))
+            ref byte encodingMap = ref s_encodingMap[0];
+
+            if (srcLength <= MaximumEncodeLength && destLength >= GetMaxEncodedToUtf8Length(srcLength))
```
Comment (jkotas) [inline src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: Would it be better to instead cap the `srcLength` to whatever fits here instead of having two copies of the encoding loop? It would avoid the penalty for the case where the output buffer is not big enough.

File: src/System.Memory/src/System/Buffers/Binary/Reader.cs [hunk 8]
```diff
@@ -106,12 +106,12 @@ public static T ReadMachineEndian<T>(ReadOnlySpan<byte> buffer)
 #else
             if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
             {
-                throw new ArgumentException(SR.Format(SR.Argument_InvalidTypeWithPointersNotSupported, typeof(T)));
+                ThrowHelper.ThrowArgumentException_InvalidTypeWithPointersNotSupported(typeof(T));
```
Comment (jkotas) [inline src/System.Memory/src/System/Buffers/Binary/Reader.cs]: What is the reason for this change?

File: src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+            int srcLength = bytes.Length;
+            int destLength = utf8.Length;
+
+            int sourceIndex = 0;
+            int destIndex = 0;
+            int result = 0;
+
+            ref byte encodingMap = ref s_encodingMap[0];
+
+            if (srcLength <= MaximumEncodeLength && destLength >= GetMaxEncodedToUtf8Length(srcLength))
```
Comment (jkotas) [inline src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: Would it make sense to cap the srcLength to whatever fits instead of having two copies of the encoding loop? It would avoid the penalty for the case where the destination buffer does not fit, and make the code quite a bit smaller (note that you get pretty non-trivial code expansion here because of the use of the AggressiveInlining attributes).

Comment (jkotas) [reply src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: Same in the decoder.

File: src/System.Memory/src/System/Buffers/Binary/Reader.cs (hunk 8 above)
Comment (ahsonkhan) [reply src/System.Memory/src/System/Buffers/Binary/Reader.cs]: From @KrzysztofCwalina 
> If you do decide to throw, throw from a helper to allow inlining.

Previously, ThrowHelper.cs was not included for netcoreapp, but I updated the project file to always include ThrowHelper.cs

Comment (jkotas) [reply src/System.Memory/src/System/Buffers/Binary/Reader.cs]: The changes in this file should have zero effect on the inlining or the native code that will be generated for this.

File: src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs (hunk 9 above)
Comment (ahsonkhan) [reply src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: Great idea. I was tinkering with that but was previously struggling to get the "cap" set correctly. But I believe I have figured it out now!

File: src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void WriteThreeLowOrderBytes(ref byte destination, int value)
+        {
+            destination = (byte)(value >> 16);
+            Unsafe.Add(ref destination, 1) = (byte)(value >> 8);
+            Unsafe.Add(ref destination, 2) = (byte)value;
+        }
+
+        // Pre-computing this table using a custom string(s_characters) and GenerateDecodingMapAndVerify (found in tests)
+        static readonly sbyte[] s_decodingMap = {
```
Comment (justinvp) [inline src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs]: `private`. corefx coding style is to be explicit with visibility.

File: src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+            int i = (oneByte << 8);
+
+            int i0 = Unsafe.Add(ref encodingMap, i >> 10);
+            int i1 = Unsafe.Add(ref encodingMap, (i >> 4) & 0x3F);
+
+            return i0 | (i1 << 8) | (EncodingPad << 16) | (EncodingPad << 24);
+        }
+
+        // Pre-computing this table using a custom string(s_characters) and GenerateEncodingMapAndVerify (found in tests)
+        static readonly byte[] s_encodingMap = {
```
Comment (justinvp) [inline src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: `private`

**Comment by ahsonkhan** [comment]
@dotnet-bot test OSX x64 Debug Build
